firedrake/preconditioners/bddc.py

```diff
@@ -0,0 +1,124 @@
+from firedrake.preconditioners.base import PCBase
+from firedrake.preconditioners.patch import bcdofs
+from firedrake.preconditioners.facet_split import restrict, get_restriction_indices
+from firedrake.petsc import PETSc
+from firedrake.dmhooks import get_function_space, get_appctx
+from firedrake.ufl_expr import TestFunction, TrialFunction
+from firedrake.functionspace import FunctionSpace, VectorFunctionSpace, TensorFunctionSpace
+from ufl import curl, div, HCurl, HDiv, inner, dx
+from pyop2.utils import as_tuple
+import numpy
+
+__all__ = ("BDDCPC",)
+
+
+class BDDCPC(PCBase):
+    """PC for PETSc PCBDDC (Balancing Domain Decomposition by Constraints).
```

Comment:
Where's the maths? What options control this PC?

Reply:
Points the docstring should cover from meeting

1. What options are - state that petsc pc has full option set description?
2. Any options added that aren't covered by the PETSc PC?
3. Are there any `appctx` entries that this PC expects or can optionally take?